users/luescher/setups/hybrid/gmm_system.py

```diff
@@ -0,0 +1,845 @@
+__all__ = ["GmmSystem"]
+
+import itertools
+import sys
+from typing import Dict
+
+# -------------------- Sisyphus --------------------
+
+from sisyphus import *
+import sisyphus.toolkit as tk
+import sisyphus.global_settings as gs
+
+# -------------------- Recipes --------------------
+
+import i6_core.lexicon.allophones as allophones
+import i6_core.am as am
+import i6_core.corpus as corpus_recipes
+import i6_core.features as features
+import i6_core.lda as lda
+import i6_core.meta as meta
+import i6_core.sat as sat
+import i6_core.rasr as rasr
+import i6_core.util as util
+import i6_core.vtln as vtln
+
+from .util import GmmDataInput, GmmPipelineArgs
+
+# -------------------- Init --------------------
+
+Path = tk.setup_path(__package__)
+
+
+# -------------------- System --------------------
+
+
+class GmmSystem(meta.System):
+    """
+    - 3 corpora types: train, dev and test
+    - only train corpora will be aligned
+    - dev corpora for tuning
+    - test corpora for final eval
+
+    to create beforehand:
+    - corpora: name and i6_core.meta.system.Corpus
+    - lexicon
+    - lm
+
+    settings needed:
+    - am
+    - lm
+    - lexicon
+    - feature extraction
+    - linear alignement
+    - monophone training
+    - monophone recognition
+    - triphone training
+    - triphone recognition
+    - vtln training
+    - vtln recognition
+    - sat training
+    - sat recognition
+    - vtln+sat training
+    - vtln+sat recognition
+    """
+
+    def __init__(self):
+        super().__init__()
+
+        self.crp["base"].python_home = gs.RASR_PYTHON_HOME
+        self.crp["base"].python_program_name = gs.RASR_PYTHON_EXE
+
+        self.gmm_args = None
+
+        self.train_corpora = []
+        self.dev_corpora = []
+        self.test_corpora = []
+
+        self.corpora = {}
+        self.concurrent = {}
+
+        self.cart_questions = None
+        self.lda_matrices = {}
+        self.cart_trees = {}
+
+        self.vtln_files = {}
+
+        self.default_align_keep_values = {
+            "default": 5,
+            "selected": gs.JOB_DEFAULT_KEEP_VALUE,
+        }
+
+    # -------------------- Setup --------------------
+    def init_system(
+        self,
+        gmm_args: GmmPipelineArgs,
+        train_data: Dict[str, GmmDataInput],
+        dev_data: Dict[str, GmmDataInput],
+        test_data: Dict[str, GmmDataInput],
+    ):
+        """
+        :param gmm_args: parameters for the different Gmm-HMM steps
+        :param train_data: dict(str: GmmInput)
+        :param dev_data: dict(str: GmmInput)
+        :param test_data: dict(str: GmmInput)
+        :return:
+        """
+        self.gmm_args = gmm_args
+        self._init_am(**gmm_args.am_args)
+        for name, v in sorted(train_data.items()):
+            self.corpora[name] = v.corpus_object
```

Comment:
This could be a separate function which could also work as external call to add corpora that are not yet in train or test.

so a `def add_corpus(name: str, corpus: meta.system.Corpus, add_lm: bool)`